modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/persistence/Database.java

```diff
@@ -348,16 +375,273 @@ public long countEvents(final String organization) {
-      final Tuple result = AssetDtos.baseJoin(em)
-          .where(QAssetDto.assetDto.checksum.eq(checksum)
-              .and(QAssetDto.assetDto.storageId.eq(storeId))
-              .and(QAssetDto.assetDto.snapshot.organizationId.eq(orgId)))
-          .singleResult(Full.select);
-      var dtoOpt = Opt.nul(result).map(Full.fromTuple);
-      return dtoOpt.isSome() ? Optional.of(dtoOpt.get()) : Optional.empty();
+      return namedQuery.findOpt(
+              "Asset.findByChecksumStorageIdAndOrganizationId",
+              AssetDto.class,
+              Pair.of("checksum", checksum),
+              Pair.of("storageId", storeId),
+              Pair.of("orgId", orgId))
+          .apply(em)
+          .map(result -> {
+            SnapshotDto snapshot = result.getSnapshot();
+            return new AssetDtos.Full(
+                result,
+                Availability.valueOf(snapshot.getAvailability()),
+                snapshot.getStorageId(),
+                snapshot.getOrganizationId(),
+                snapshot.getOwner(),
+                snapshot.getMediaPackageId(),
+                snapshot.getVersion()
+            );
+          });
+    });
+  }
+
+  public Optional<Snapshot> getLatestSnapshot(String mediaPackageId) {
+    return getLatestSnapshot(mediaPackageId, null);
+  }
+
+  public Optional<Snapshot> getLatestSnapshot(String mediaPackageId, String orgId) {
+    return db.execTx(em -> {
+      Optional<SnapshotDto> snapshotDto = namedQuery.findOpt(
+          "Snapshot.findLatest",
+          SnapshotDto.class,
+          Pair.of("mediaPackageId", mediaPackageId),
+          Pair.of("organizationId", orgId)
+      ).apply(em);
+
+      if (snapshotDto.isEmpty()) {
+        return Optional.empty();
+      }
+
+      Snapshot snapshot = snapshotDto.get().toSnapshot();
+      // make sure the delivered media package has valid URIs
+      snapshot = httpAssetProvider.prepareForDelivery(snapshot);
+
+      return Optional.of(snapshot);
+    });
+  }
+
+  public Optional<MediaPackage> getMediaPackage(String mediaPackageId) {
+    return getMediaPackage(mediaPackageId, null);
+  }
+  public Optional<MediaPackage> getMediaPackage(String mediaPackageId, String orgId) {
+    return getLatestSnapshot(mediaPackageId, orgId).map(Snapshot::getMediaPackage);
+  }
+
+  public List<Snapshot> getSnapshots(String mediaPackageId) {
+    return getSnapshots(mediaPackageId, null);
+  }
+
+  public List<Snapshot> getSnapshots(String mediaPackageId, String orgId) {
+    return getSnapshots(mediaPackageId, orgId, null);
+  }
+
+  public List<Snapshot> getSnapshots(String mediaPackageId, String orgId, String orderByVersion) {
+    return db.execTx(em -> {
+      String namedQueryName = "Snapshot.findByMpId";
+      if ("ASC".equals(orderByVersion)) {
+        namedQueryName = "Snapshot.findOldestVersionFirst";
+      }
+      if ("DESC".equals(orderByVersion)) {
+        namedQueryName = "Snapshot.findLatestVersionFirst";
+      }
```

Comment:
We never promised that queries are unsorted if you don't pass any sort parameter → you can have one named query less:

```suggestion
      // Optional: check parameter
      if (!List.of("ASC", "DESC").contains(orderByVersion)) {
        // Exception here
      }

      String namedQueryName = "ASC".equals(orderByVersion) ? "Snapshot.findOldestVersionFirst" : "Snapshot.findLatestVersionFirst";
```